VmAgent.Core/ProcessOutputLogger.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Azure.Gaming.VmAgent.Core
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.IO;
+    using ApplicationInsights;
+    using ApplicationInsights.DataContracts;
+    using global::VmAgent.Core.Dependencies.Interfaces.Exceptions;
+    using Microsoft.Azure.Gaming.VmAgent.Core.Dependencies.Interfaces;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    public class ProcessOutputLogger: IProcessOutputLogger
+    {
+        private string _logFilePath;
+        private StreamWriter _logWriter;
+        private MultiLogger _logger;
+
+        public ProcessOutputLogger(string logFilePath, MultiLogger logger)
+        {
+            if (String.IsNullOrEmpty(logFilePath))
+            {
+                throw new ProcessOuputLoggerCreationFailedException($"LogFilePath cannot be null or empty.");
+            }
+
+            try
+            {
+                _logWriter = new StreamWriter(File.OpenWrite(logFilePath));
+            }
+            catch (Exception ex)
+            {
+                _logger.LogInformation($"Exception was thrown while creating a Process Log file under. {ex}");
+                throw new ProcessOuputLoggerCreationFailedException($"Process Output Logger failed to create a file. {ex}.");
+            }
+
+            _logFilePath = logFilePath;
+            _logger = logger;
+
+            _logger.LogInformation($"Process Log file is created under {_logFilePath}");
+        }
+
+        public void Log(string message)
+        {
+            if (_logWriter == null)
+            {
+                _logger.LogInformation($"StreamWriter is null or not created yet.");
+                return;
+            }
+
+            string currentDate = DateTime.Now.ToString("yyyy-MM-dd'T'HH:mm:ss.fffK");
+
+            _logWriter.WriteLine($"{currentDate}\t{message}");
+            _logWriter.Flush();
+        }
+
+        public string GetProcessLogFilePath()
+        {
+            return _logFilePath;
+        }
+
+        public void Close()
```

Comment:
is close something that you want people to be able to call explicitly? If they have a closed object, should they be able to interact with it in any way? 

To me, this seems like a very standard Disposable/Using pattern, where you create something for a given scope and then it should go away outside of that scope. 

If that's not the case, this probably should have some kind of state that tracks whether or not the object is closed and throws an exception if they attempt to use it when it's closed. As far as I can tell if this was closed you would get a log saying 'streamwriter is null or not created yet' in the Log call, but you wouldn't get any error condition.

Reply:
Thanks Kalon, good questions - My initial though was file should be open whne VmAgent starts running a new Process and file should not be exposed to users. File would be closed when Process is terminated because we need to capture all the output and errors of Process while it is running. Only Logger interacts with Process through event handlers. Do you think I missed something here?

Reply:
So there are a couple possibilities.

One is that Process implements IDisposable and has a disposable ProcessOutputLogger as part of its objects, and when it is disposed of it disposes of this object. That's probably the standard way of chaining disposable objects together. However, since ProcessRunner isn't disposable that's probably not what you want. 

So the next thing to do is make sure that this class enforces how it is supposed to act - which means that anyone who uses this class gets an error if they attempt to log if it is closed. What that means is either you have a separate variable to track if it is closed, or you use logWriter as a proxy for that, and whenever logwriter is null and you attempt to log you actually throw an exception instead of just logging. 

Another possible issue is that calling Close() is not always the right choice due to how the filesystem works - and if multiple things are logging/writing at once it might cause an issue. Generally you should call Dispose() instead of Close when dealing with streams, because they will do things like suppress finalizers and deal with other race conditions.

Reply:
@KalonJ-MSFT  Thank you for the feedback - I will think about the way to track the opened files using LogWriter. I think we can keep the list of LogWriter and check regularly. For possible issue you mentioned,  I thought close() eventually calls dispose() in FileSystem - I will double check with that too.

Reply:
I think you're right and I misremebered - Close() calls Dispose():

[https://stackoverflow.com/questions/7524903/should-i-call-close-or-dispose-for-stream-objects](https://stackoverflow.com/questions/7524903/should-i-call-close-or-dispose-for-stream-objects)

So close is the right thing and I was wrong. Sorry!

Reply:
I made an update so that it can throw an exception instead of just logging - exceptions would be thrown in wrapper class so it will not break any code. I was thinking we could manage the each filesystem object and monitor them like session hosts but it will be trickier and I don't come up with the cases when file is still open or alive after the process (game session) is terminated because file instance would be alive in the same scope as each Process where event handler is triggered from.

Reply:
I think this is a good place to end it. if it becomes a problem later we can revisit it then.